torch/csrc/api/include/torch/optim/lbfgs.h

```diff
@@ -14,59 +14,83 @@
-      : LossClosureOptimizer(std::forward<ParameterContainer>(parameters)),
-        options(options_),
-        ro(options_.history_size()),
-        al(options_.history_size()) {}
+struct TORCH_API LBFGSParamState : public OptimizerCloneableParamState<LBFGSParamState> {
+  TORCH_ARG(int64_t, func_evals) = 0;
+  TORCH_ARG(int64_t, n_iter) = 0;
+  TORCH_ARG(double, t);
+  TORCH_ARG(Tensor, d) = {};
+  TORCH_ARG(Tensor, H_diag) = {};
+  TORCH_ARG(Tensor, prev_flat_grad) = {};
+  TORCH_ARG(Tensor, prev_loss) = {};
+  TORCH_ARG(std::deque<Tensor>, old_dirs);
+  TORCH_ARG(std::deque<Tensor>, old_stps);
+  TORCH_ARG(std::deque<Tensor>, ro);
+  TORCH_ARG(std::vector<Tensor>, al);
 
-  torch::Tensor step(LossClosure closure) override;
+ public:
+  void serialize(torch::serialize::InputArchive& archive) override;
+  void serialize(torch::serialize::OutputArchive& archive) const override;
+  TORCH_API friend bool operator==(const LBFGSParamState& lhs, const LBFGSParamState& rhs);
+  ~LBFGSParamState() = default;
+};
 
-  LBFGSOptions options;
+class TORCH_API LBFGS : public Optimizer {
+ public:
+   explicit LBFGS(std::vector<OptimizerParamGroup> param_groups,
+       LBFGSOptions defaults) : Optimizer(std::move(param_groups), std::make_unique<LBFGSOptions>(defaults)) {
+     TORCH_CHECK(param_groups_.size() == 1, "LBFGS doesn't support per-parameter options (parameter groups)");
+     if (defaults.max_eval() == c10::nullopt) {
+       auto max_eval_val = (defaults.max_iter() * 5) / 4;
+       auto& options = static_cast<LBFGSOptions&>(param_groups_[0].options());
+       options.max_eval(max_eval_val);
+       auto& default_options = static_cast<LBFGSOptions&>(*defaults_.get());
+       default_options.max_eval(max_eval_val);
+     }
+     _params = &param_groups_[0].params();
+     _numel_cache = c10::nullopt;
+   }
+   explicit LBFGS(
+       std::vector<Tensor> params,
+       LBFGSOptions defaults) : LBFGS({std::move(OptimizerParamGroup(params))}, defaults) {}
 
+  Tensor step(LossClosure closure) override;
+  void add_parameters(const std::vector<Tensor>& parameters) override;
+  const std::vector<Tensor>& parameters() const noexcept override;
+  std::vector<Tensor>& parameters() noexcept override;
+  size_t size() const noexcept override;
   void save(serialize::OutputArchive& archive) const override;
   void load(serialize::InputArchive& archive) override;
 
-  Tensor d{torch::empty({0})};
-  Tensor H_diag{torch::empty({0})};
-  Tensor prev_flat_grad{torch::empty({0})};
-  Tensor t{torch::zeros(1)};
-  Tensor prev_loss{torch::zeros(1)};
-  std::vector<Tensor> ro;
-  std::vector<Tensor> al;
-  std::deque<Tensor> old_dirs;
-  std::deque<Tensor> old_stps;
-  int64_t func_evals{0};
-  int64_t state_n_iter{0};
-
  private:
-  LBFGS() : options(0) {}
-
-  Tensor gather_flat_grad();
-  void add_grad(const torch::Tensor& step_size, const Tensor& update);
+  std::vector<Tensor>* _params;
```

Comment:
hmm why are we storing a pointer to `std::vector<Tensor>` as a field of this class? This can potentially lead to segfaults if the object being pointed to is out of scope (i.e. erased from memory), since pointer to an object doesn't take any ownership of the object and cannot guarantee object lifetime.

I believe we can use `std::vector<Tensor>` here and it will do what you want it to do - changing value of a tensor in `self.param_groups[0]['params']` will also change its value in this vector, since the tensors in these two containers still share the same TensorImpl.

Reply:
No but storing it as a vector creates problems when a new tensor is added. sure we can add them to both param_groups_[0] and _params but it seems suboptimal to do so

Reply:
also we would have to update both _params and param_groups_[0] in functions like _add_grad and set_param where params are being updated in place.

Reply:
We could also decide to get rid of _params completely and just use param_groups_[0].params() instead

Reply:
> also we would have to update both _params and param_groups_[0].params() in functions like _add_grad and set_param where params are being updated in place.

I believe for operations like `p.add_()` and `_params->at(i).copy_()`, updating the tensors in `_params` will update `param_groups_[0].params()` at the same time, because those tensors share the same TensorImpl, and doing in-place update on one side will also update the other side.

Reply:
> No but storing it as a vector creates problems when a new tensor is added. sure we can add them to both param_groups_[0] and _params but it seems suboptimal to do so

Curious how is this currently being handled in the Python implementation? My understanding (might be wrong) is that Python doesn't fully support adding a new Tensor into `param_groups[0].params` either, and it would be great to make C++/Python implementation be the same (which is making `_params` a `std::vector<Tensor>` and assign `param_groups_[0].params()` to it in the LBFGS constructor).

Reply:
> We could also decide to get rid of _params completely and just use param_groups_[0].params() instead

Discussed offline: I think you are right as Python uses reference semantics for assignment, and we should get rid of `_params` and just use `param_groups_[0].params()` instead :)